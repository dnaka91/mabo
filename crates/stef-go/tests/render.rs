use std::{
    fmt::Write,
    fs,
    path::{Path, PathBuf},
};

use insta::{assert_snapshot, glob, with_settings};
use stef_go::{Opts, Output};
use stef_parser::Schema;

fn strip_path(path: &Path) -> PathBuf {
    path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs"))
        .or_else(|_| path.strip_prefix(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/inputs_extra")))
        .unwrap()
        .to_owned()
}

fn merge_output(buf: &mut String, output: Output<'_>, parent: &Path) {
    let path = parent.join(output.name);
    let _ = write!(buf, "--- {}.go\n\n{}", path.display(), output.content);

    for module in output.modules {
        merge_output(buf, module, &path);
    }
}

#[test]
fn render_schema() {
    glob!("inputs/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str(), Some(&strip_path(path))).unwrap();
        let value = stef_go::render_schema(&Opts { package: "sample" }, &value);

        let mut merged = String::new();
        merge_output(&mut merged, value, Path::new(""));

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("render", merged);
        });
    });
}
