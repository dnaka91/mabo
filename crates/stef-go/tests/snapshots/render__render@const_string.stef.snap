---
source: crates/stef-go/tests/render.rs
description: "const SIMPLE: string = \"value\";\n\nconst NEWLINE_ESCAPE: string = \"one \\\n    two \\\n    three\\\n\";\n\nconst ESCAPES: string = \"escape basics \\r\\n \\t \\b \\f \\\\ \\\"\\\n    hello\\\" \\n\\\n    unicode \\u{2764}  \\\n    emoji ❤ \\\n\";\n\nconst MULTILINE: string = \"a\n    b\n    c\n\";"
input_file: crates/stef-parser/tests/inputs/const_string.stef
---
--- sample.go

// Code generated by stef-go (v0.1.0). DO NOT EDIT.

import (
	"github.com/dnaka91/stef-go"
	"github.com/dnaka91/stef-go/buf"
)

package sample

const Simple string = "value"
const NewlineEscape string = "one two three"
const Escapes string = "escape basics \r\n \t \u{8} \u{c} \\ \"hello\" \nunicode ❤  emoji ❤ "
const Multiline string = "a\n    b\n    c\n"

